File: blink-macos/BlinkCore/Tests/BlinkCoreTests/BreakComplianceTrackerTests.swift
import Foundation
import Testing
@testable import BlinkCore

@Suite("BreakComplianceTracker")
struct BreakComplianceTrackerTests {
    @Test("Break taken within 60s = taken")
    func breakTaken() {
        let tracker = BreakComplianceTracker()
        var recorded: BreakRecord?
        tracker.onBreakRecorded = { recorded = $0 }

        let promptTime = Date()
        tracker.breakPrompted(at: promptTime, flowState: .normal, flowScore: 0.5)

        tracker.breakTaken(at: promptTime.addingTimeInterval(10), idleDuration: 20)

        #expect(recorded?.compliance == .taken)
    }

    @Test("Break dismissed quickly = dismissed")
    func breakDismissed() {
        let tracker = BreakComplianceTracker()
        var recorded: BreakRecord?
        tracker.onBreakRecorded = { recorded = $0 }

        let promptTime = Date()
        tracker.breakPrompted(at: promptTime, flowState: .normal, flowScore: 0.5)

        tracker.breakDismissed(at: promptTime.addingTimeInterval(2))

        #expect(recorded?.compliance == .dismissed)
    }

    @Test("Break dismissed after delay = delayed")
    func breakDelayed() {
        let tracker = BreakComplianceTracker()
        var recorded: BreakRecord?
        tracker.onBreakRecorded = { recorded = $0 }

        let promptTime = Date()
        tracker.breakPrompted(at: promptTime, flowState: .flow, flowScore: 0.8)

        tracker.breakDismissed(at: promptTime.addingTimeInterval(30))

        #expect(recorded?.compliance == .delayed)
    }

    @Test("Break ignored = ignored")
    func breakIgnored() {
        let tracker = BreakComplianceTracker()
        var recorded: BreakRecord?
        tracker.onBreakRecorded = { recorded = $0 }

        let promptTime = Date()
        tracker.breakPrompted(at: promptTime, flowState: .normal, flowScore: 0.3)

        tracker.breakIgnored(at: promptTime.addingTimeInterval(300))

        #expect(recorded?.compliance == .ignored)
        #expect(recorded?.respondedAt == nil)
    }

    @Test("No prompt = no record")
    func noPrompt() {
        let tracker = BreakComplianceTracker()
        var recorded: BreakRecord?
        tracker.onBreakRecorded = { recorded = $0 }

        tracker.breakTaken(at: Date(), idleDuration: 20)
        #expect(recorded == nil, "Should not record without a prompt")
    }

    @Test("isTrackingBreak reflects state")
    func trackingState() {
        let tracker = BreakComplianceTracker()
        #expect(!tracker.isTrackingBreak)

        tracker.breakPrompted(at: Date(), flowState: .normal, flowScore: 0.5)
        #expect(tracker.isTrackingBreak)

        tracker.breakTaken(at: Date(), idleDuration: 20)
        #expect(!tracker.isTrackingBreak)
    }

    @Test("Flow state and score are captured at prompt time")
    func capturesPromptState() {
        let tracker = BreakComplianceTracker()
        var recorded: BreakRecord?
        tracker.onBreakRecorded = { recorded = $0 }

        tracker.breakPrompted(at: Date(), flowState: .deepFlow, flowScore: 0.92)
        tracker.breakTaken(at: Date(), idleDuration: 20)

        #expect(recorded?.flowStateWhenPrompted == .deepFlow)
        #expect(recorded?.flowScore == 0.92)
    }
}

File: blink-windows/src/Blink.Core.Tests/BreakComplianceTrackerTests.cs
using Blink.Core.Compliance;
using Blink.Core.FlowDetection;

namespace Blink.Core.Tests;

public class BreakComplianceTrackerTests
{
    [Fact]
    public void BreakTaken_Within60s()
    {
        var tracker = new BreakComplianceTracker();
        BreakRecord? recorded = null;
        tracker.OnBreakRecorded += r => recorded = r;

        var prompt = DateTime.Now;
        tracker.BreakPrompted(prompt, FlowState.Normal, 0.5);
        tracker.BreakTaken(prompt.AddSeconds(10), 20);

        Assert.NotNull(recorded);
        Assert.Equal(BreakCompliance.Taken, recorded.Compliance);
    }

    [Fact]
    public void BreakDismissed_Quickly()
    {
        var tracker = new BreakComplianceTracker();
        BreakRecord? recorded = null;
        tracker.OnBreakRecorded += r => recorded = r;

        var prompt = DateTime.Now;
        tracker.BreakPrompted(prompt, FlowState.Normal, 0.5);
        tracker.BreakDismissed(prompt.AddSeconds(2));

        Assert.Equal(BreakCompliance.Dismissed, recorded!.Compliance);
    }

    [Fact]
    public void BreakDismissed_AfterDelay()
    {
        var tracker = new BreakComplianceTracker();
        BreakRecord? recorded = null;
        tracker.OnBreakRecorded += r => recorded = r;

        var prompt = DateTime.Now;
        tracker.BreakPrompted(prompt, FlowState.Flow, 0.8);
        tracker.BreakDismissed(prompt.AddSeconds(30));

        Assert.Equal(BreakCompliance.Delayed, recorded!.Compliance);
    }

    [Fact]
    public void BreakIgnored()
    {
        var tracker = new BreakComplianceTracker();
        BreakRecord? recorded = null;
        tracker.OnBreakRecorded += r => recorded = r;

        var prompt = DateTime.Now;
        tracker.BreakPrompted(prompt, FlowState.Normal, 0.3);
        tracker.BreakIgnored(prompt.AddSeconds(300));

        Assert.Equal(BreakCompliance.Ignored, recorded!.Compliance);
        Assert.Null(recorded.RespondedAt);
    }

    [Fact]
    public void NoPrompt_NoRecord()
    {
        var tracker = new BreakComplianceTracker();
        BreakRecord? recorded = null;
        tracker.OnBreakRecorded += r => recorded = r;

        tracker.BreakTaken(DateTime.Now, 20);
        Assert.Null(recorded);
    }

    [Fact]
    public void IsTrackingBreak_ReflectsState()
    {
        var tracker = new BreakComplianceTracker();
        Assert.False(tracker.IsTrackingBreak);

        tracker.BreakPrompted(DateTime.Now, FlowState.Normal, 0.5);
        Assert.True(tracker.IsTrackingBreak);

        tracker.BreakTaken(DateTime.Now, 20);
        Assert.False(tracker.IsTrackingBreak);
    }

    [Fact]
    public void CapturesPromptState()
    {
        var tracker = new BreakComplianceTracker();
        BreakRecord? recorded = null;
        tracker.OnBreakRecorded += r => recorded = r;

        tracker.BreakPrompted(DateTime.Now, FlowState.DeepFlow, 0.92);
        tracker.BreakTaken(DateTime.Now, 20);

        Assert.Equal(FlowState.DeepFlow, recorded!.FlowStateWhenPrompted);
        Assert.Equal(0.92, recorded.FlowScore);
    }
}
